### DISCUSSION TITLE
Stuttering Width-animation

### DISCUSSION THREAD

**Discussion by llfab** [AvaloniaUI/Avalonia#9736] [Q&A] [answered]
Because I noticed in my application that uses a sliding menu that the translational animations seem not so smooth, I created a very simple width animation on a Grid without content. Also there, I find it non-smooth. I keep comparing this with translational animations that used in Qt Qml that also uses an OpenGL backend. They are super smooth even on a much slower computer. 

Hence, I thought I open a discussion on whether or not the animation should be smoother. 

Details:
- Windows 10 21H2
- 11th Gen Intel(R) Core(TM) i7-1165G7 @ 2.80GHz  with Intel Iris Xe
- Avalonia 11.0.0-preview4

Code:
```csharp
<Window xmlns="https://github.com/avaloniaui"
        xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
        xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
        xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
        mc:Ignorable="d" d:DesignWidth="800" d:DesignHeight="450"
        x:Class="AvaloniaApplication1.MainWindow"
        Title="AvaloniaApplication1">
    <Window.Styles>
        <Style Selector="Grid.animGrid">
            <Setter Property="Transitions">
                <Transitions>
                    <DoubleTransition Property="Width" Duration="0:0:2.4"/>
                </Transitions>
            </Setter>
            <Setter Property="Background" Value="Green"/>
            <Setter Property="Width" Value="200"/>
        </Style>
    </Window.Styles>
    <StackPanel Orientation="Vertical">
        <TextBox Text="Text Box"></TextBox>
        <CheckBox IsChecked="True">CheckBox</CheckBox>
        <Button >Button</Button>
        <ToggleButton Checked="OnChecked" Unchecked="OnUnchecked">Toggle</ToggleButton>
        <Grid x:Name="_animGrid" Classes="animGrid" HorizontalAlignment="Left" Height="200">
        </Grid>
    </StackPanel>
</Window>
```

Code behind ```OnChecked``` and ```OnUnchecked``` change the width.

See the animation in this video:
https://user-images.githubusercontent.com/31031996/208297511-d3921f77-c6ad-4e3b-ba83-367eb29e9dfe.mp4

I have also tried it on another machine with the same CPU but Nvidia Quadro T2000 and it is not really better there.

Should we expect something smoother or do you guys find that good enough?

Note: on Opacity animations you do not notice such stuttering but that may have to do with the difference in optical impression (Opacity vs. Translation).

**Comment by kekekeks** [comment]
Animating properties that trigger a layout pass currently has performance issues.
In general for sliding menus it's recommended to animate RenderTransform instead of layout

**Reply by llfab** [reply]
Thanks for the quick reply. I tried the same with scaleX RenderTransform and I do not see a difference.

https://user-images.githubusercontent.com/31031996/208298460-e92f6210-417f-48f1-9cd2-9bebd1db1d51.mp4

```csharp
<Window xmlns="https://github.com/avaloniaui"
        xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
        xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
        xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
        mc:Ignorable="d" d:DesignWidth="800" d:DesignHeight="450"
        x:Class="AvaloniaApplication1.MainWindow"
        Title="AvaloniaApplication1">
    <Window.Styles>
        <Style Selector="Grid.animGrid">
            <Setter Property="Transitions">
                <Transitions>
                    <DoubleTransition Property="Width" Duration="0:0:2.4"/>
                    <DoubleTransition Property="Height" Duration="0:0:2.4"/>
                    <TransformOperationsTransition Property="RenderTransform" Duration="0:0:2.4"/>
                </Transitions>
            </Setter>
            <Setter Property="Background" Value="Green"/>
            <Setter Property="Width" Value="200"/>
            <Setter Property="RenderTransformOrigin" Value="0.0,0.0"/>
            <Setter Property="RenderTransform" Value="scaleX(1)"/>

            <Style Selector="^.large">
                <Setter Property="RenderTransform" Value="scaleX(3)" />
            </Style>
        </Style>
    </Window.Styles>
    <StackPanel Orientation="Vertical">
        <TextBox Text="Text Box"></TextBox>
        <CheckBox IsChecked="True">CheckBox</CheckBox>
        <Button >Button</Button>
        <ToggleButton Checked="OnChecked" Unchecked="OnUnchecked">Toggle</ToggleButton>
        <Grid x:Name="_animGrid" Classes="animGrid" HorizontalAlignment="Left" Height="200">
        </Grid>
    </StackPanel>
</Window>```

Classes is added/removed with ```large``` in code behind

**Reply by maxkatz6** [reply]
@llfab I have the same results as you when I disable UseWindowsUIComposition option. Can you confirm that winui composition is actually used in your test app? You can find that by checking instance type of `AvaloniaLocator.Current.GetService<IRenderTimer>()` result.

Also, please keep in mind that old Vulkan PR wasn't using proper render timer either as per @kekekeks. You can't get smooth FPS this way. P.S. Updated Vulkan implementation is coming after #9639 was merged.

**Reply by llfab** [reply]
How is this?

![grafik](https://user-images.githubusercontent.com/31031996/208303561-47b7c850-a7fa-43f3-bf7d-220215aae49d.png)

**Reply by llfab** [reply]
Clarifying: I am using the 0.10.18 template, having updated the nuget packages to 11.0.0-preview4. The builder code is:
```csharp
    internal class Program
    {
        // Initialization code. Don't use any Avalonia, third-party APIs or any
        // SynchronizationContext-reliant code before AppMain is called: things aren't initialized
        // yet and stuff might break.
        [STAThread]
        public static void Main(string[] args) => BuildAvaloniaApp()
            .StartWithClassicDesktopLifetime(args);

        // Avalonia configuration, don't remove; also used by visual designer.
        public static AppBuilder BuildAvaloniaApp()
            => AppBuilder.Configure<App>()
                .UsePlatformDetect()
                .LogToTrace();
    }

```

That is ok, right?

**Comment by llfab** [comment]
Well, if you compare it very often you may have the impression that it is less stuttering but still not really smooth.

**Comment by maxkatz6** [comment, marked as answer]
It's a problem of layout animations in all of XAML based framework that use same measure+arrange passes argorithms. On each animation tick whole sub tree needs to be recalculated to get new layout properties.

Theoretically, there is not much space for improvements, we can only add more micro-optimizations in this area.
And developers can replace layout animations with render animations where it's possible.

But practically, seeing that you have problem with empty Grid layout, there could be another problem. Possibly not related with animations at all. Ignoring the fact that Grid by itself is the most complex panel in the framework with heaviest layout passes (it's empty, shouldn't matter).

Upd: while writing this, noticed that you have the same problem with render-animations. Which hints you might have a problem with render timer, or something related to rendering.

**Reply by maxkatz6** [reply]
> Note: on Opacity animations you do not notice such stuttering but that may have to do with the difference in optical impression (Opacity vs. Translation).

If it was only about layout animations, opacity transitions don't trigger layout recalculations.

**Reply by llfab** [reply]
Got it.

I also replaced the animations in my complex slide menu with render anims and that turned out to not improve things as well. So I thought to report that. I will create videos for the slide menu to let you decide whether that is something that should be tackled.

**Reply by maxkatz6** [reply]
@llfab see this comment, must be a reason of animation problems - https://github.com/AvaloniaUI/Avalonia/discussions/9736#discussioncomment-4441857

**Comment by llfab** [comment]
I have changed the Grid to a Panel with the same results and checked CompositionRenderer vs. NonCompositionRenderer as follows:

Non Composition:
https://user-images.githubusercontent.com/31031996/208307095-09fdef82-6745-477b-8f99-e9c0e3986f51.mp4

Composition:
https://user-images.githubusercontent.com/31031996/208307108-8457c87f-3ef0-46dd-b879-2ee97ffee8c0.mp4

There is a visible difference but with CompositionRenderer it is still not very smooth.

I am using this startup code:
```
        public static AppBuilder BuildAvaloniaApp()
            => AppBuilder.Configure<App>()
                .UsePlatformDetect()
                .With(new Win32PlatformOptions() { UseWindowsUIComposition = true })
                .LogToTrace();
```
... toggling true and false.

**Reply by llfab** [reply]
So it is smoother but still has some "hops".

**Reply by llfab** [reply]
Just to clarify: The original videos from the start of this discussion were all done with activated Windows UI Composition. So the comparisson here just shows that it get worse without that renderer...

**Comment by llfab** [comment]
@maxkatz6 : I have played more:
- it turns out that you can get your machine into a mode where graphics is not at 100% especially when switch from docking station to integrated monitor
- I was able to verify that above simple sample works much smoother indeed when using render transform. 
- If you disable compositional rendering even with render transform it will be less smooth. 
- Hence, the reported problem from above may have more to do with non optomal GPU situation
- I still have the feeling that Qt Qml which is also scene graph based is much smoother here, however, that has also some 10 more years of improvements in his back pack

**Comment by llfab** [comment]
Here is an example of our slide menu which currently walks through all measure/arrange passes and can be optimized using render transforms.
https://user-images.githubusercontent.com/31031996/208403442-04745297-04c8-48f9-a740-e8274fef306c.mp4
